### ISSUE TITLE
Interacting with JavaScript, DOM and Web APIs?

### ISSUE THREAD

**Issue by trueadm** [evanw/zeroscript#9]
Are there any plans to support interaction with JavaScript, DOM and Web APIs with ThinScript? I'd be really interested in seeing the performance benefit that low-level WebAssembly can bring when creating large web apps. I'd also be interested in exposing API end-points for other JavaScript libraries to consume/interact with the compiled application.

I haven't got a huge experience in this field and I'm not even sure it might all be possible/feasible, but I thought it might be a good discussion point.

**Comment by evanw** [comment]
Not exactly. I have plans to support general-purpose JavaScript function imports, so you'll definitely be able to use DOM APIs in that sense. WebAssembly doesn't provide any way of bypassing the JavaScript layer and talking directly to the DOM at the moment so it's likely that you won't see any performance improvement with code where the DOM is a bottleneck. WebAssembly can only do what JavaScript can already do. You're much more likely to see performance improvements in DOM-heavy code by changing your code so it makes less use of the DOM. [React](https://facebook.github.io/react/) or other [virtual DOM](https://github.com/Matt-Esch/virtual-dom) approaches are good examples of this.

In the far future, the WebAssembly project has indicated that it would like to investigate maybe including support for accessing garbage-collected objects from WebAssembly itself. This will not be part of the initial MVP and won't be coming any time soon. If that does happen, it would be amazing if JavaScript JIT compiler engineers could add optimizations that inline C++ function calls directly into the generated assembly code. From what I understand, there is currently some inefficiency with JavaScript/C++ bindings because of extra argument validation that is required every time the boundary is crossed. Inlining C++ function calls would remove all of that overhead by allowing the JIT optimizer to remove unnecessary validation checks. That would require a lot of changes to WebAssembly and JavaScript JIT design though and it's all just pure speculation from me in response to your question.

**Comment by trueadm** [comment]
@evanw 

> You're much more likely to see performance improvements in DOM-heavy code by changing your code so it makes less use of the DOM. React or other virtual DOM approaches are good examples of this.

I'm already author of [Inferno](https://github.com/trueadm/inferno), one of the fastest virtual DOM libraries around. I was hoping I could get more benefits if I could interact with the DOM API directly without going through the JavaScript layer, but as you say, this isn't likely to be possible.

You're doing a great job here and I'll definitely keep an eye on this project and your work :)

**Comment by evanw** [comment]
I just came across some interesting thoughts on DOM overhead here from Boris Zbarsky (Mozilla): https://news.ycombinator.com/item?id=11780060.
